pkg/controller/license/trial/trial_controller_test.go

```diff
@@ -0,0 +1,208 @@
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+const testNs = "ns-1"
+const trialSecretName = "eck-trial" // user's choice but constant simplifies test setup
+
+var trialSecretNsn = types.NamespacedName{
+	Namespace: testNs,
+	Name:      trialSecretName,
+}
+
+func trialSecretSample(annotated bool, data map[string][]byte) *v1.Secret {
+	sec := v1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      trialSecretName,
+			Namespace: testNs,
+			Labels: map[string]string{
+				licensing.LicenseLabelType: "enterprise-trial", // assume always there otherwise watch would not trigger
+			},
+			Annotations: map[string]string{},
+		},
+		Data: data,
+	}
+	if annotated {
+		sec.Annotations[licensing.EULAAnnotation] = licensing.EULAAcceptedValue
+	}
+	return &sec
+}
+
+func trialLicenseBytes() []byte {
+	return []byte(fmt.Sprintf(
+		`{"license": {"uid": "x", "type": "enterprise-trial", "issue_date_in_millis": 1, "expiry_date_in_millis": %d, "issued_to": "x", "issuer": "x", "start_date_in_millis": 1, "cluster_licenses": null, "Version": 0}}`,
+		chrono.ToMillis(time.Now().Add(24*time.Hour)), // simulate a license still valid for 24 hours
+	))
+
+}
+
+func testPubkey(t *testing.T) *rsa.PublicKey {
+	rnd := rand.Reader
+	key, err := rsa.GenerateKey(rnd, 2048)
+	require.NoError(t, err)
+	return &key.PublicKey
+}
+
+func testPubkeyBytes(t *testing.T) []byte {
+	bytes, err := x509.MarshalPKIXPublicKey(testPubkey(t))
+	require.NoError(t, err)
+	return bytes
+}
+
+func TestReconcileTrials_Reconcile(t *testing.T) {
+	requireValidationMsg := func(msg string) func(c k8s.Client) {
+		return func(c k8s.Client) {
+			var sec v1.Secret
+			require.NoError(t, c.Get(trialSecretNsn, &sec))
+			err, ok := sec.Annotations[licensing.LicenseInvalidAnnotation]
+			require.True(t, ok, "invalid annotation present")
+			require.Equal(t, msg, err)
+		}
+	}
+
+	type fields struct {
+		Client      k8s.Client
+		trialPubKey *rsa.PublicKey
+	}
+	tests := []struct {
+		name       string
+		fields     fields
+		wantErr    bool
+		assertions func(c k8s.Client)
+	}{
+		{
+			name: "trial secret needs accepted EULA",
+			fields: fields{
+				Client: k8s.WrappedFakeClient(trialSecretSample(false, nil)),
+			},
+			wantErr:    false,
+			assertions: requireValidationMsg(EULAValidationMsg),
+		},
+		{
+			name: "valid trial secret inits trial",
+			fields: fields{
+				Client: k8s.WrappedFakeClient(trialSecretSample(true, nil)),
+			},
+			wantErr: false,
+			assertions: func(c k8s.Client) {
+				var sec v1.Secret
+				require.NoError(t, c.Get(types.NamespacedName{
+					Namespace: testNs,
+					Name:      licensing.TrialStatusSecretKey,
+				}, &sec))
+				require.NoError(t, c.Get(trialSecretNsn, &sec))
+				_, lic, err := licensing.TrialLicense(c, trialSecretNsn)
+				require.NoError(t, err)
+				require.NoError(t, lic.IsMissingFields())
+			},
+		},
+		{
+			name: "valid trial after operator restart",
```

Comment:
I'm not sure I understand this test, the name and the assertion seems a little bit confusing to me. IIUC the license check is failing here because there is no valid license in `trialLicenseBytes` , not because it is "restarted" ?

Reply:
Yes I am not happy with this either. Let me think if there is a better way to test this.

Reply:
I re-did the test, with a valid simulated trial.